stdlib/src/collections/_unicode.mojo

```diff
@@ -0,0 +1,265 @@
+        var index = _lowercase_mapping_index(rune)
+        if index != -1:
+            found = True
+            continue
+        index = _uppercase_mapping_index(rune)
+        if index != -1:
+            return False
+        index = _uppercase_mapping2_index(rune)
+        if index != -1:
+            return False
+        index = _uppercase_mapping3_index(rune)
+        if index != -1:
+            return False
+    return found
+
+
+fn is_lowercase(s: String) -> Bool:
+    """Returns True if all characters in the string are lowercase, and
+        there is at least one cased character.
+
+    Args:
+        s: The string to examine.
+
+    Returns:
+        True if all characters in the string are lowercase and
+        there is at least one cased character, False otherwise.
+    """
+    var found = False
+    for c in s:
+        var rune = ord(c)
+        var index = _uppercase_mapping_index(rune)
+        if index != -1:
+            found = True
+            continue
+        index = _uppercase_mapping2_index(rune)
+        if index != -1:
+            found = True
+            continue
+        index = _uppercase_mapping3_index(rune)
+        if index != -1:
+            found = True
+            continue
+        index = _lowercase_mapping_index(rune)
+        if index != -1:
+            return False
+    return found
+
+
+fn _ord(_p: UnsafePointer[UInt8]) -> (Int, Int):
+    """Return the rune and number of bytes to be consumed, for given UTF-8 string pointer
+    """
+    var p = _p
+    var b1 = p[]
+    if (b1 >> 7) == 0:  # This is 1 byte ASCII char
+        return int(b1), 1
+    var num_bytes = count_leading_zeros(~b1)
+    var shift = int((6 * (num_bytes - 1)))
+    var b1_mask = 0b11111111 >> (num_bytes + 1)
+    var result = int(b1 & b1_mask) << shift
+    for _ in range(1, num_bytes):
+        p += 1
+        shift -= 6
+        result |= int(p[] & 0b00111111) << shift
+    return result, int(num_bytes)
+
+
+fn _write_rune(rune: UInt32, p: UnsafePointer[UInt8]) -> Int:
+    """Write rune as UTF-8 into provided pointer. Return number of added bytes.
+    """
+    if (rune >> 7) == 0:  # This is 1 byte ASCII char
+        p[0] = rune.cast[DType.uint8]()
+        return 1
+
+    @always_inline
+    fn _utf8_len(val: UInt32) -> Int:
+        alias sizes = SIMD[DType.uint32, 4](
+            0, 0b1111_111, 0b1111_1111_111, 0b1111_1111_1111_1111
+        )
+        var values = SIMD[DType.uint32, 4](val)
+        var mask = values > sizes
+        return int(mask.cast[DType.uint8]().reduce_add())
+
+    var num_bytes = _utf8_len(rune)
+    var shift = 6 * (num_bytes - 1)
+    var mask = UInt32(0xFF) >> (num_bytes + 1)
+    var num_bytes_marker = UInt32(0xFF) << (8 - num_bytes)
+    p[0] = (((rune >> shift) & mask) | num_bytes_marker).cast[DType.uint8]()
+    for i in range(1, num_bytes):
+        shift -= 6
+        p[i] = (((rune >> shift) & 0b00111111) | 0b10000000).cast[DType.uint8]()
+    return num_bytes
```

Comment:
This code logic is used in many other places. I'll open a PR to fix `chr` implementation which I put in #3239 and ended up mixed with other problems and getting forgotten.